File: content/snippets/typescript/interview_lcci/04/01_route_between_nodes.ts
function hasRoute(n: number, edges: Array<[number, number]>, start: number, target: number): boolean {
  const g: number[][] = Array.from({ length: n }, () => []);
  for (const [u, v] of edges) g[u].push(v);
  const seen = Array(n).fill(false);
  const q: number[] = [start];
  seen[start] = true;
  while (q.length) {
    const u = q.shift()!;
    if (u === target) return true;
    for (const v of g[u]) {
      if (!seen[v]) {
        seen[v] = true;
        q.push(v);
      }
    }
  }
  return false;
}

const edges: Array<[number, number]> = [[0, 1], [0, 2], [1, 3], [2, 4]];
console.log(hasRoute(5, edges, 0, 4));
console.log(hasRoute(5, edges, 3, 4));

File: content/snippets/cpp/interview_lcci/04/01_route_between_nodes.cpp
#include <iostream>
#include <queue>
#include <vector>

bool hasRoute(int n, const std::vector<std::pair<int, int>>& edges, int start, int target) {
    std::vector<std::vector<int>> g(n);
    for (auto [u, v] : edges) g[u].push_back(v);
    std::vector<int> seen(n, 0);
    std::queue<int> q;
    q.push(start);
    seen[start] = 1;
    while (!q.empty()) {
        int u = q.front(); q.pop();
        if (u == target) return true;
        for (int v : g[u]) if (!seen[v]) { seen[v] = 1; q.push(v); }
    }
    return false;
}

int main() {
    std::vector<std::pair<int, int>> edges{{0,1},{0,2},{1,3},{2,4}};
    std::cout << std::boolalpha << hasRoute(5, edges, 0, 4) << '\n';
    std::cout << std::boolalpha << hasRoute(5, edges, 3, 4) << '\n';
}
